### PR TITLE
test(ScoreAgent): test done for the scores hehe

### LINKED ISSUE #37: test score_agent

### PR DESCRIPTION
**Pull request by Xiaorui-Huang**
### Motivation and context

Closes: #37 
### Before submitting this PR, please make sure:
 - [x] Your code builds cleanly without any errors

### CODE REVIEW

**Comment by Xiaorui-Huang** [comment]
added test and some preconditions, there is a question for the tests tho if anyone knows the answer, setting up private variables without setters is a pain in the ass

File: woodoku/entity/score_agent.py [hunk 1]
```diff
@@ -26,6 +26,9 @@ def calculate_winning(self, blocks: int, groups: int) -> None:
             blocks (int): The number of blocks placed
             groups (int): The number of groups completed
         """
+        # Note: assertion can be turned of with -O flag to python
```
Comment (YANLING-H) [inline woodoku/entity/score_agent.py]:
```suggestion
        # Note: assertion can be turned off with -O flag to python
```

File: woodoku/entity/score_agent.py [hunk 2]
```diff
@@ -26,6 +26,9 @@ def calculate_winning(self, blocks: int, groups: int) -> None:
...
+        assert blocks > 0 and groups >= 0, "Precondition violated"
+
```
Comment (YANLING-H) [inline woodoku/entity/score_agent.py]: I understand that the assertion can be turned off when running the final version of the code in the release branch, but should we remove all assertions after finishing training models? I just want to bring up the discussion here. It makes sense for me to keep all assertions now since some wired edge cases might be encountered later when training machine learning models.

File: woodoku/tests/entity/test_score_agent.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+def fixture_scorekeeper() -> ScoreAgent:
+    return ScoreAgent()
+
+
+# Question... (just wanna know if there is a better way)
+# Is there any way to use Mock or MagicMock to set the private state instead of
+# a hack like this or accessing the private variable
+# for example: I want to set up so that before testing the scorekeeper has a
+# score of 42 and current streak of 1
+@pytest.fixture(name="setup_and_apply_winnings", autouse=True)
```
Comment (YANLING-H) [inline woodoku/tests/entity/test_score_agent.py:20]: I did find a way to mock property [in MagicMock](https://stackoverflow.com/questions/55512708/how-to-properly-mock-private-members-of-a-class) but I think mocking is not applicable in testing scoreAgent without changing any method in scoreAgent since once we make a scoreAgent a mock object, we can no longer call calculate_winning which is the function to test.

File: woodoku/tests/entity/test_score_agent.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+    score_before_streak = (
+        init_score
+        # remove score added by blocks (init_streak * 1)
+        - init_streak
+        # remove score added by groups (init_streak * GROUP_POINTS)
+        - init_streak * GROUP_POINTS
+        # remove scores added by streaks (0 + 10 + 20... for a total init_streak
+        # number of times i.e. [(n-1) * n]/2)
+        - init_streak * (init_streak - 1) // 2 * STREAK_POINTS
+    )
```
Comment (YANLING-H) [inline woodoku/tests/entity/test_score_agent.py:60]: Could you please explain a little more about score_before_streak? I didn't follow.

File: woodoku/entity/score_agent.py (hunk 2 above)
Comment (Xiaorui-Huang) [reply woodoku/entity/score_agent.py]: The interpreter simply skips any assertion in the production/release version. There is no need to delete it.

File: woodoku/tests/entity/test_score_agent.py (hunk 4 above)
Comment (Xiaorui-Huang) [reply woodoku/tests/entity/test_score_agent.py:60]: It's a little too complicated to explain if you can't make out the meaning from the code. The goal is to set up the score and the streak to the desired inputs, and as long as the assertion passes the code should be fine.

There is a better way to set up using mock that I have not implemented yet...

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -20,6 +22,6 @@ repos:
         ]
     - id: mypy 
       name: mypy 
-      entry: mypy
-      language: python
-      types: [python]
\ No newline at end of file
+      entry: pipenv run mypy
+      language: system
+      types: [python]
```
Comment (EdwardHaoranLee) [inline .pre-commit-config.yaml:27]: What's wrong with the version without pipenv?

File: woodoku/tests/entity/test_score_agent.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+@pytest.fixture(name="setup_and_apply_winnings", autouse=True)
+def fixture_setup_and_apply_winnings(
+    init_score: int,
+    init_streak: int,
+    winnings: Iterable[Tuple[int, int]],
+    scorekeeper: ScoreAgent,
+) -> None:
+    """
+    Automatically setup up scorekeeper and apply the winning to scorekeeper
+    for any test with with fixture/parameters
```
Comment (EdwardHaoranLee) [inline woodoku/tests/entity/test_score_agent.py]:
```suggestion
    for any test with fixture/parameters
```

File: woodoku/tests/entity/test_score_agent.py (hunk 4 above)
Comment (EdwardHaoranLee) [reply woodoku/tests/entity/test_score_agent.py:60]: Yep, may consider use mocking instead.

File: woodoku/tests/entity/test_score_agent.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+    # elif init_score > 0 and not init_streak:
+    # scorekeeper.calculate_winning(init_score, 0)
+    assert init_score == scorekeeper.get_score()
+
+    # apply the winnings
+    for winning in winnings:
+        scorekeeper.calculate_winning(*winning)
+
+
+class TestScoreAgent:
```
Comment (EdwardHaoranLee) [inline woodoku/tests/entity/test_score_agent.py:73]: Really good test cases.

**Review by EdwardHaoranLee — DISMISSED** [review]

File: .pre-commit-config.yaml (hunk 5 above)
Comment (Xiaorui-Huang) [reply .pre-commit-config.yaml:27]: This is more of an issue regarding environments... Normally from the terminal, you would be in some virtual environment that has all the dependencies installed, but using a UI tool uses the system context instead of the virtual environment. This is what happens with vscode version control UI, so the pipenv tell it to pre-commit using the pipenv virtual environment

File: woodoku/tests/entity/test_score_agent.py (hunk 4 above)
Comment (Xiaorui-Huang) [reply woodoku/tests/entity/test_score_agent.py:60]: I've done a little bit more investigation. I think mocking here isn't the best approach either... What I intended to mock was the get function so that I can control the output of the getters within the `calculate_winning` functions. But this then means that the test setup is dependent on the implementation of `calculate_winning`. Because what I really want to set up is the private variables and not some functions, Mocking here is also just a hack... 

Mocking as a white box approach here seems inferior to a black box approach, where I can ignore how `calculate_winning` is implemented and use it to set up the tests assuming it's correct based solely on its inputs and what it does (the scoring rules). And if the setup fails it means that `calculate winning` was not implemented correctly. which is also a test in itself.

Sticking with the current setup abstracts the implementation of `calculate winning` ... the only downside is there is a little bit of math involved to reserve engineer what scores need to be added before setting up streaks (since streak also have addition point)

**Review by EdwardHaoranLee — DISMISSED** [review]

**Review by EdwardHaoranLee — APPROVED** [review]
